### PR TITLE
feat: new highlight creation popup (with highlight suggestions)

### LINKED ISSUE #943: Feature: Highlight suggestions
### Type of feature

🍕 Feature

### Current behavior

We need to provide things for the user to highlight. 

### Suggested solution

Can we design the follow
1. button on the contribution list that suggests creating a highlight
2. Pull Request suggestions in a list next to the feed suggesting a highlight. 

### Additional context

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### Contributing Docs

- [ ] I agree to follow this project's Contribution Docs

Image

### PR DESCRIPTION
**Pull request by diivi**
## Description
#943 
PR for testing/QA

## What type of PR is this? (check all applicable)

- [x] 🍕 Feature
- [ ] 🐛 Bug Fix
- [ ] 📝 Documentation Update
- [ ] 🎨 Style
- [ ] 🧑‍💻 Code Refactor
- [ ] 🔥 Performance Improvements
- [ ] ✅ Test
- [ ] 🤖 Build
- [ ] 🔁 CI
- [ ] 📦 Chore (Release)
- [ ] ⏩ Revert

## Related Tickets & Documents
Closes #943 

## Added tests?

- [ ] 👍 yes
- [x] 🙅 no, because they aren't needed
- [ ] 🙋 no, because I need help

## Added to documentation?

- [ ] 📜 README.md
- [ ] 📓 docs.opensauced.pizza
- [ ] 🍕 dev.to/opensauced
- [ ] 📕 storybook
- [x] 🙅 no documentation needed

### CODE REVIEW

**Comment by diivi** [comment]
Current bugs I'm investigating:

1. Whenever I change the body or call any `setState` hook like `setHighlightLink`, the list of suggestion disappears :/
2. calling handleGenerateHighlightSummary` raises the "invalid URL..." error even with the correct URL, probably another state related issue.

**Comment by diivi** [comment]
Remaining:

1. Tooltips appear behind the dialog currently, need to bring them up.
2. I'll see what to do about the mobile form.
3. Carousel to show all 9 suggestions that i fetched.

**Comment by bdougie** [comment]
Some quick feedback 

We shouldn't recommend highlight for issues or prs that are already highlighted by the user. My most popular issue is being recommend even though I already have a highlight for it.

**Comment by diivi** [comment]
forgot this, will do.

**Comment by diivi** [comment]
image
image
image

**Comment by diivi** [comment]
I can't reproduce the netlify errors locally 🤔

**Comment by bdougie** [comment]
> I can't reproduce the netlify errors locally 🤔

sharing for context

```
Type error: Type '{ children: Element[]; spaceBetween: number; slidesPerView: number; className: string; modules: any[]; pagination: { clickable: boolean; }; a11y: { enabled: boolean; }; }' is not assignable to type 'IntrinsicAttributes & RefAttributes<SwiperRef> & SwiperProps'.
1:43:29 PM:   Property 'spaceBetween' does not exist on type 'IntrinsicAttributes & RefAttributes<SwiperRef> & SwiperProps'.
1:43:29 PM:   682 | 
1:43:29 PM:   683 |             <Swiper
1:43:29 PM: > 684 |               spaceBetween={8}
1:43:29 PM:       |               ^
1:43:29 PM:   685 |               slidesPerView={1}
1:43:29 PM:   686 |               className=lg:w-[33vw] md:w-[50vw]
1:43:29 PM:   687 |               modules={[Pagination, A11y]}
```

**Comment by diivi** [comment]
It says "neutral" now after merging with the beta branch.

File: components/molecules/HighlightInput/highlight-input-form.tsx
Comment (babblebey) [inline components/molecules/HighlightInput/highlight-input-form.tsx]: I thought It would be nice to have this function here return a filtered array of suggestions that hasn't been highlighted yet.

This IMO will help reduce the friction of setting the state of `highlightSuggestions` twice... as seen in the next comment..

File: components/molecules/HighlightInput/highlight-input-form.tsx
Comment (babblebey) [inline components/molecules/HighlightInput/highlight-input-form.tsx]:
```js
await removeAlreadyHighlightedSuggestions(newHighlightSuggestions); // <- setHighlightSuggestions(newHighlightSuggestions)
setHighlightSuggestions(newHighlightSuggestions); // setting the state again.
```

The `await removeAlreadyHighlightedSuggestions(newHighlightSuggestions);` function already sets the state of the `highlightSuggestions` within, but you are once again setting the state in next line `setHighlightSuggestions(newHighlightSuggestions);`

If you let the `removeAlreadyHighlightedSuggestions` return a filtered array of un-highlighted suggestions, you will have something some readable like so...

```js
const filteredSuggestions = await removeAlreadyHighlightedSuggestions(newHighlightSuggestions);
setHighlightSuggestions(filteredSuggestions);
```

File: components/molecules/HighlightInput/highlight-input-form.tsx
Comment (diivi) [reply components/molecules/HighlightInput/highlight-input-form.tsx]: Thanks for pointing this out, missed it while iterating.

**Comment by isabensusan** [comment]
Hey diivi! First of all, thank you so much for your great work :)

Here's a couple style fixes to make the interface look even more polished.

## Text fixes

**Purple arrows**
There's multiple different text styles, sizes and colors in the modal, we should normalize these.
The purple arrows highlight the text that needs changing. The tailwind classes should all be:
- [ ] size: .text-sm 
- [ ] weight: .font-normal
- [ ] color for the text: .text-light-slate-12  (so when users write on the textarea and input components, it should be this color)
- [ ] color for the placeholder text:  .text-light-slate-9

<br/>

**Red arrows**
The titles of the suggested highlights should stay almost the same styling-wise, but:
- [ ] let's make them links to the github issue/PR
- [ ] on-hover the text should be orange

<br/>

**Green arrow**
- [ ] There's a trailing space in the placeholder text that adds a bit of padding making the input component look unbalanced. Let's remove that

image





## General Style Fixes
**Orange**
- [ ] Make paddings .p2 : The padding for the input forms is different on each component. Let's make them all the same 

<br/>

**Blue**
- [ ] Remove shadow on this input component so it looks like the other input forms

<br/>

**Yellow**
- [ ] Make the highlight suggestion's top padding .py-1

<br/>

**Teal**
- [ ] Make both icons 20x20 (right now the sparkle is 16x16, make them both the same)
- [ ] Reduce gap between both icons to .gap-0.5

<br/>

**Pink**
- [ ] Make the slide indicators bigger, they are 6x6 right now, let's change that to 8x8

image


## "Add a repo" component fixes
This component is looking a bit unbalanced, let's try these changes:
**Orange** (placeholder text)
- [ ] Change font classes to these: .text-sm .text-light-slate-9 .font-normal .pl-2 .pr-2

<br/>

**Blue**
- [ ] change "Add a repo" button text size to .text-xs to match the tag component 

image

<br/>

**This is how we want it to look:**
image

**Comment by isabensusan** [comment]
Also, a couple more things from doing QA:

## Issue 1
It takes some time to load the highlights, and the empty state in the meantime looks broken. Can we add a loading skeleton animation?

**Something like this, but animated:**
image

**Gif Example:**
image


## Issue 2
When I click on the edit button for a highlight suggestion, the image for the highlight is way to big.
Kapture 2023-09-04 at 21 04 14

**Comment by diivi** [comment]
Thanks for the extensive design review! 

>  color for the placeholder text: .text-light-slate-9

This wasn't giving consistent results for the placeholders, so I went ahead with text-slate-400.

Different looking placeholder text colors for the same `text-light-slate-9` class:
image

**Comment by isabensusan** [comment]
> This wasn't giving consistent results for the placeholders, so I went ahead with text-slate-400.

This is ok, as long as all placeholders in the modal (for the highlights textarea, for the repo tag input and for the auto-summarize component) are the same color, font-size and font-weight.

Let me know when you are ready for another review! Thanks :)

**Comment by diivi** [comment]
It's ready for a review, I think I covered all points.

**Comment by OgDev-01** [comment]
> I think I covered all points

Great Job @diivi ... 

It would also be nice if we could take off the title for the highlights as requested in #1660

File: components/molecules/HighlightInput/highlight-input-form.tsx
Comment (jpmcb) [inline components/molecules/HighlightInput/highlight-input-form.tsx:855]: I get the same Netlify error on running `npm run prebuild && npm run build`:

```
info  - Need to disable some ESLint rules? Learn more here: https://nextjs.org/docs/basic-features/eslint#disabling-rules
- info Linting and checking validity of types ...Failed to compile.

./components/molecules/HighlightInput/highlight-input-form.tsx:627:17
Type error: Type '{ children: Element[]; spaceBetween: number; slidesPerView: number; className: string; modules: any[]; pagination: { clickable: boolean; }; a11y: { enabled: boolean; }; }' is not assignable to type 'IntrinsicAttributes & RefAttributes<SwiperRef> & SwiperProps'.
  Property 'spaceBetween' does not exist on type 'IntrinsicAttributes & RefAttributes<SwiperRef> & SwiperProps'.

  625 |             ) : (
  626 |               <Swiper
> 627 |                 spaceBetween={8}
      |                 ^
  628 |                 slidesPerView={1}
  629 |                 className="lg:w-[33vw] md:w-[50vw]"
  630 |                 modules={[Pagination, A11y]}
```

My node environment:
```
❯ npm --version
npm info using npm@9.6.7
npm info using node@v18.17.1
9.6.7
npm info ok
```

I do notice after running those build steps I get the following diff in the shrinkwrap which might be irrelivant:
```diff
diff --git a/npm-shrinkwrap.json b/npm-shrinkwrap.json
index 1bcc45a..d155c00 100644
--- a/npm-shrinkwrap.json
+++ b/npm-shrinkwrap.json
@@ -118,8 +118,8 @@
         "typesync": "^0.10.0"
       },
       "engines": {
-        "node": ">= 16.7.0",
-        "npm": ">= 8.0.0"
+        "node": "^18.17.0 || >=20.5.0",
+        "npm": ">= 9.6.0"
       }
     },
     "node_modules/@adobe/css-tools": {
```

Looks like there are similar errors people report with using an older version of TypeScript:
https://github.com/nolimits4web/swiper/issues/6843#issuecomment-1634818157

I'm guessing it would be a pretty huge lift to upgrade our typescript dependency:
https://github.com/open-sauced/app/blob/b1782cb094ff687296324f3354704c6681520d44/package.json#L152

**Review by jpmcb — CHANGES_REQUESTED** [review]

File: components/molecules/HighlightInput/highlight-input-form.tsx
Comment (brandonroberts) [reply components/molecules/HighlightInput/highlight-input-form.tsx:855]: We should be able to upgrade the TypeScript dependency without much of a lift in separate PR.

**Comment by isabensusan** [comment]
Design review approved! thanks @diivi great work 💪

File: components/molecules/HighlightInput/highlight-input-form.tsx
Comment (jpmcb) [reply components/molecules/HighlightInput/highlight-input-form.tsx:858]: Looks like the last deploy still fails with this warning:
https://app.netlify.com/sites/oss-insights/deploys/64fb63ae7fde590008a744ae

Should we try upgrading TS as part of this PR if we need the swiper component?

**Comment by brandonroberts** [comment]
@diivi what's left to do here besides fixing the build errors?

**Comment by diivi** [comment]
> @diivi what's left to do here besides fixing the build errors?

nothing, I'm waiting for reviews.

**Review by brandonroberts — APPROVED** [review]
Works! 🍕

**Review by isabensusan — APPROVED** [review]

**Review by jpmcb — APPROVED** [review]
🕺🏼

**Pull request merged**
